### DISCUSSION TITLE
tracking observables within async getters

### DISCUSSION THREAD

**Discussion by farcaller** [mobxjs/mobx#2894] [Q&A]
TL;DR: [here's the sandbox](https://codesandbox.io/s/mobx-scribble-eval-sync-hpxjt?file=/src/store.js) and I'm trying to make the `Scribble.JSMoudule()` into an async function (similar to this broken [sandbox](https://codesandbox.io/s/mobx-scribble-eval-async-sbd2s?file=/src/store.js)).

I'm trying to refactor my API from throwing upon failing to resolve the computed props to a more neat concept of promises and I'm thoroughly confused if I'm doing something wrong or mobx just doesn't support this (slightly insane) use case.

## My data model

I have a store similar to 

```ts
export class Scribble {
  title = "";
  body = "";

  JSModule = computedFn(function JSModule() {
    ...
  });
}

export class Store {
  scribblesByTitle = new Map();
}
```

`JSModule` does the processing over `this.body` and returns the result.

## Introducing inter-scribble dependencies

Now, part of that it might need to get the body of another `Scribble` by title from the store. To facilitate that I need to pass back the reference to the store for scribbles I create:

```ts
export class Scribble {
  title = "";
  body = "";
  store;

  constructor(store) {
    makeAutoObservable(this, {
      store: false
    });
    this.store = store;
  }

  JSModule = computedFn(function JSModule() {
    return loadModule(this, {
      importScribble: (title) => {
        console.log(`importing ${title} from ${this.title}`);
        const sc = this.store.scribblesByTitle.get(title);
        return sc.JSModule();
      }
    });
  });
}
```

## Introducing partial scribbles

Due to `body` potentially being large and the number of scribbles growing, the API doesn't fetch it when it syncs to the server. Thus, the scribbles actually look like

```ts
export class Scribble {
  body: string | null;
  ...
}
```

Where `null` signifies that I need to call out to the API and fetch the body when needed.

## Current state of things

[I figured a sandbox would be the best to demonstrate](https://codesandbox.io/s/mobx-scribble-eval-sync-hpxjt?file=/src/store.js).

Because it's fully synchronous, whenever I call to `parent.JSModule()` in my react code `observer` adds the computed's dependencies and thus tracks the changes in any child:

![image](https://user-images.githubusercontent.com/693/114680407-0fc68d80-9d05-11eb-988f-f558f446fd59.png)

So if I update the child scribble, the components that depend on parent scribble re-render too.

## And this is where it all breaks apart with async

[sandbox](https://codesandbox.io/s/mobx-scribble-eval-async-sbd2s?file=/src/store.js)

Supposedly, my `importScribble` is an async function now. It calls into the store, gets the child scribble, if the child's body is `null` it calls to the API, waits for the network and then returns the resolved child's `body`. In the sample sandbox I simulate the network delay with `setTimeout`. Because I can't just return the `Promise` in the computed getter I tried making a workaround with:

```js
  // XXX: we need to keep the promise in the local state as it's no longed computed
  // and every re-render will trigger more promises and then Chrome explodes
  const [modulePromise, setModulePromise] = useState(
    Promise.resolve(undefined)
  );

  useEffect(() => {
    console.log("updating promise for", scribble.title);
    setModulePromise(scribble.JSModule());
  }, [scribble, scribble.body]); // XXX: I don't think this is correct as it's mixing
  // mobx state with react state...

  let module;
  fromPromise(modulePromise).case({
    pending: () => {
      console.log("eval pending for", scribble.title);
      module = "PENDING";
    },
    fulfilled: (v) => {
      console.log("eval fulfiled for", scribble.title, v);
      module = v;
    },
    rejected: (e) => {
      console.log("eval rejected for", scribble.title, e);
      module = `err ${e.message}`;
    }
  });
```

However this has two issues:
- it's overly clunky (heh)
- it cannot observe the dependencies that arise through the Promise border, so it cannot track the child updates in the parent (you can see that by pressing "update child" in both sandboxes. Only the sync version updates the parent)

Any advice on how to move forward with this?

**Comment by farcaller** [comment]
I think this actually dedups against #668 and #872, now that I spent even more time researching it. I guess there's no easy way out with promises.
